scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -184,98 +184,148 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     case Tokenized.Success(tokens) => tokens
     case Tokenized.Error(_, _, details) => throw details
   }
-  lazy val (parserTokens, parserTokenPositions) = {
-    val parserTokens = mutable.ArrayBuilder.make[Token]()
-    val parserTokenPositions = mutable.ArrayBuilder.make[Int]()
-    @tailrec def loop(prevPos: Int, currPos: Int, sepRegions: List[Char]): Unit = {
-      if (currPos >= scannerTokens.length) return
-      val prev = if (prevPos >= 0) scannerTokens(prevPos) else null
-      val curr = scannerTokens(currPos)
-      val nextPos = {
-        var i = currPos + 1
-        while (i < scannerTokens.length && scannerTokens(i).is[Trivia]) i += 1
-        if (i == scannerTokens.length) i = -1
-        i
-      }
-      val next = if (nextPos != -1) scannerTokens(nextPos) else null
-      // SIP-27 Trailing comma (multi-line only) support.
-      // If a comma is followed by a new line & then a closing paren, bracket or brace
-      // then it is a trailing comma and is ignored.
-      def isTrailingComma: Boolean =
-        dialect.allowTrailingCommas &&
-          curr.is[Comma] &&
-          next.is[CloseDelim] &&
-          next.pos.startLine > curr.pos.endLine
-      if (curr.isNot[Trivia] && !isTrailingComma) {
-        parserTokens += curr
-        parserTokenPositions += currPos
-        val sepRegions1 = {
-          if (curr.is[LeftParen]) ')' :: sepRegions
-          else if (curr.is[LeftBracket]) ']' :: sepRegions
-          else if (curr.is[LeftBrace]) '}' :: sepRegions
-          else if (curr.is[CaseIntro]) '\u21d2' :: sepRegions
-          else if (curr.is[RightBrace]) {
-            var sepRegions1 = sepRegions
-            while (!sepRegions1.isEmpty && sepRegions1.head != '}') sepRegions1 = sepRegions1.tail
-            if (!sepRegions1.isEmpty) sepRegions1 = sepRegions1.tail
-            sepRegions1
-          } else if (curr.is[RightBracket]) { if (!sepRegions.isEmpty && sepRegions.head == ']') sepRegions.tail else sepRegions }
-          else if (curr.is[RightParen]) { if (!sepRegions.isEmpty && sepRegions.head == ')') sepRegions.tail else sepRegions }
-          else if (curr.is[RightArrow]) { if (!sepRegions.isEmpty && sepRegions.head == '\u21d2') sepRegions.tail else sepRegions }
-          else sepRegions // do nothing for other tokens
+
+
+  trait TokenIterator extends Iterator[Token] {
+    def prevTokenPos: Int;
+
+    def tokenPos: Int;
+
+    def token: Token;
+
+    def fork: TokenIterator
+  }
+
+  var in: TokenIterator = new SimpleTokenIterator()
+
+  private class SimpleTokenIterator(var curTokenPos: Int = -1,
+                                    var prevPos: Int = -1,
+                                    var sepRegions: List[Char] = List()) extends TokenIterator {
+
+    var curToken: Token = null
+
+    if (curTokenPos == -1) fetchToken() else curToken = scannerTokens(curTokenPos)
```

Comment:
I don't believe this is the right way to do it. If you fork on a new line the value of `token` will not be the same for the current iterator and the forked one.

I suggest that `curToken` becomes a parameter of the class with a default value of `null`.
I would then revert the constructor body to what it was before:
```scala
 if (curTokenPos == -1) next() // NOTE: only do next() if we've been just created. forks can't go for next()
```
And `fork` becomes:
```scala
def fork: TokenIterator = new SimpleTokenIterator(curToken, curTokenPos, prevPos, sepRegions)
```

Also since you have
```scala
def token: Token = curToken
def prevTokenPos: Int = prevPos
def tokenPos: Int = curTokenPos
```
I would just remove these definitions and rename `curToken` to `token`, `prevPos` to `prevTokenPos` and `curTokenPos` to  `tokenPos`.

Scala allows this:
```scala
trait Foo { def foo: Int }
class FooImpl(var foo: Int) extends Foo
```